test/e2e_tests/test_convergence_wait_abort.py

```diff
@@ -0,0 +1,120 @@
+                for log_line in log_file:
+                    if keyword in log_line:
+                        return True 
+                return False 
+        
+        resource_manifest_path = os.path.join(test_data_dir, "k8s-event-watcher", "unsatisfiable-affinity.yaml")
+        self._test_convergence_wait_abort("unsatisfiable-affinity", resource_manifest_path, log_file_test)
+
+
+    def test_invalid_image(self):
+        def log_file_test(log_file_path) -> bool:
+            keyword = "Aborting convergence wait due to failed predicate (reason: Failed,"
+            with open(log_file_path, "r") as log_file:
+                for log_line in log_file:
+                    if keyword in log_line:
+                        return True 
+                return False 
+        
+        resource_manifest_path = os.path.join(test_data_dir, "k8s-event-watcher", "invalid-image.yaml")
+        self._test_convergence_wait_abort("invalid-image", resource_manifest_path, log_file_test)
+
+
+
+
+
+    # should never abort a convergence wait for satisfiable deployments
+    def test_satisfiable_deployment(self):
+        def log_file_test(log_file_path) -> bool:
+            keyword = "Aborting convergence"
+            with open(log_file_path, "r") as log_file:
+                for log_line in log_file:
+                    if keyword in log_line:
+                        return False
+                return True
+        
+        resource_manifest_path = os.path.join(test_data_dir, "k8s-event-watcher", "satisfiable-deployment.yaml")
+        self._test_convergence_wait_abort("satisfiable", resource_manifest_path, log_file_test)
+
+    
+    
+    # apply a resource manifest and examine the log file
+    def _test_convergence_wait_abort(self, cluster_name:str, resource_file_path: str, log_test_predicate: Callable[[str],bool]) -> str:
+        
+        tmp_dir = tempfile.TemporaryDirectory()
+
+        
+        log_file_path = os.path.join(tmp_dir.name, "test.log")
+
+        logging.basicConfig(
+            filename=log_file_path,
+            level=logging.WARN,
+            format="%(message)s",
+            force=True
+        )
+
+        kube_config_path = os.path.join(os.path.expanduser("~"), ".kube/test-"+cluster_name)
+        
+        cluster = Kind(
+            acto_namespace=0, num_nodes=3, version="v1.27.3"
+        )
+
+        cluster.create_cluster(cluster_name, kube_config_path)
+
+
+        runner = Runner(
+            context = {
+                "namespace": "test",
+                "crd": None,
+                "preload_images": set(),
+            },
+            trial_dir=tmp_dir.name,
+            kubeconfig= kube_config_path,
+            context_name="kind-"+cluster_name
+        )
+
+        utils.create_namespace(runner.apiclient, "test") 
+
+        runner.run_without_collect(resource_file_path)
+        cluster.delete_cluster(cluster_name, kube_config_path)
+
+        assert(log_test_predicate(log_file_path))
```

Comment:
Test to see if an abort is triggered by examining the log file, since I haven't been able to find a cleaner way to test that.